### PR TITLE
Document System.IO.WindowsRuntimeStorageExtensions

### PR DESCRIPTION
**Pull request by carlossanlop**
@JeremyKuhne wrote these. Can you please take a look?

API Proposal: https://github.com/dotnet/corefx/issues/18759
PR: https://github.com/dotnet/corefx/pull/19029
Public information in StackOverflow by MSFT member about the new Windows interfaces: https://stackoverflow.com/questions/42799235/how-can-i-get-a-win32-handle-for-a-storagefile-or-storagefolder-in-uwp#42799236

Files:
https://github.com/dotnet/corefx/blob/master/src/System.Runtime.WindowsRuntime/src/System/IO/WindowsRuntimeStorageExtensions.cs#L179

The public API for IStorageFolderHandleAccess.Create: https://docs.microsoft.com/en-us/windows/win32/api/windowsstoragecom/nf-windowsstoragecom-istoragefolderhandleaccess-create
The public API for IStorageItemHandleAccess.Create: https://docs.microsoft.com/en-us/windows/win32/api/windowsstoragecom/nf-windowsstoragecom-istorageitemhandleaccess-create

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
https://github.com/dotnet/dotnet-api-docs/issues/1962 has the original request.

File: xml/System.IO/WindowsRuntimeStorageExtensions.xml [hunk 1]
```diff
@@ -52,6 +52,28 @@ The next example shows the XAML code that is associated with the previous exampl
     </remarks>
   </Docs>
   <Members>
+    <MemberGroup MemberName="CreateSafeFileHandle">
+      <AssemblyInfo>
+        <AssemblyName>System.Runtime.WindowsRuntime</AssemblyName>
+        <AssemblyVersion>4.0.14.0</AssemblyVersion>
+      </AssemblyInfo>
+      <Docs>
+        <summary>Creates and returns a safe file handle for the IStorage instance that is being extended.</summary>
```
Comment (JeremyKuhne) [inline xml/System.IO/WindowsRuntimeStorageExtensions.xml]: This should be a reference to SafeFileHandle

File: xml/System.IO/WindowsRuntimeStorageExtensions.xml [hunk 2]
```diff
@@ -52,6 +52,28 @@ The next example shows the XAML code that is associated with the previous exampl
...
+        <remarks>
+          <format type="text/markdown">
+            <![CDATA[
+
```
Comment (JeremyKuhne) [inline xml/System.IO/WindowsRuntimeStorageExtensions.xml]: The key thing here is that you can get a brokered SafeFileHandle. This is critically important if you want to create a FileStream around an IStorageItem or IStorageFolder when you are running in an [AppContainer](https://docs.microsoft.com/en-us/windows/win32/secauthz/appcontainer-isolation). When running in an AppContainer the application has very limited file access rights. Things like picture and document folders require the native file handle to be proxied via RuntimeBroker.exe as the application process itself doesn't have rights to most files. This API gets the proxied handle. If accessing user data folders or using a file picker, this API should always be used to create `FileStream`. Using a path will often throw `AccessDeniedException` in these cases.

@mairaw I don't know how to write this in a more user friendly way. :)

Comment (JeremyKuhne) [reply xml/System.IO/WindowsRuntimeStorageExtensions.xml]: Fyi: Universal Windows apps run in an AppContainer.

File: xml/System.IO/WindowsRuntimeStorageExtensions.xml (hunk 1 above)
Comment (carlossanlop) [reply xml/System.IO/WindowsRuntimeStorageExtensions.xml]:
```suggestion
        <summary>Creates and returns a reference to a safe file handle for the IStorage instance that is being extended.</summary>
```

File: xml/System.IO/WindowsRuntimeStorageExtensions.xml (hunk 2 above)
Comment (carlossanlop) [reply xml/System.IO/WindowsRuntimeStorageExtensions.xml]: Thanks for your suggested remark, @JeremyKuhne . Do you want me to substitute the original remark with yours, or do you want me to append it?

File: xml/System.IO/WindowsRuntimeStorageExtensions.xml [hunk 3]
```diff
@@ -52,6 +52,28 @@ The next example shows the XAML code that is associated with the previous exampl
...
+## Remarks
+
+With the Windows 10 Anniversary Update, new interfaces were added to <xref:Windows.Storage.IStorageFolder> and <xref:IStorageFile> that allow creating a standard Win32 file handle: `IStorageFolderHandleAccess` and `IStorageItemHandleAccess`. The `CreateSafeFileHandle` group of extension methods take advantage of the `Create` methods of these interfaces.
+
+Getting the Win32 handle allows richer interoperability with both .NET and other existing libraries. You can, for example, go straight to a <xref:System.IO.FileStream> without complicated and potentially performance-sapping wrappers.
+
```
Comment (carlossanlop) [inline xml/System.IO/WindowsRuntimeStorageExtensions.xml]:
```suggestion
The key thing here is that you can get a brokered `SafeFileHandle`. This is critically important if you want to create a <xref:System.IO.FileStream> around an <xref:Windows.Storage.IStorageItem> or <xref:Windows.Storage.IStorageFolder> when you are running in an [AppContainer](https://docs.microsoft.com/en-us/windows/win32/secauthz/appcontainer-isolation). When running in an AppContainer, the application has very limited file access rights. Things like picture and document folders require the native file handle to be proxied via `RuntimeBroker.exe` as the application process itself doesn't have rights to most files. This API gets the proxied handle. If accessing user data folders or using a file picker, this API should always be used to create <xref:System.IO.FileStream>. Using a path will often throw <xref:System.AccessDeniedException> in these cases.

```

File: xml/System.IO/WindowsRuntimeStorageExtensions.xml [hunk 4]
```diff
@@ -52,6 +52,28 @@ The next example shows the XAML code that is associated with the previous exampl
...
+      </AssemblyInfo>
+      <Docs>
+        <summary>Creates and returns a safe file handle for the IStorage instance that is being extended.</summary>
+        <remarks>
+          <format type="text/markdown">
+            <![CDATA[
+
+## Remarks
+
+With the Windows 10 Anniversary Update, new interfaces were added to <xref:Windows.Storage.IStorageFolder> and <xref:IStorageFile> that allow creating a standard Win32 file handle: `IStorageFolderHandleAccess` and `IStorageItemHandleAccess`. The `CreateSafeFileHandle` group of extension methods take advantage of the `Create` methods of these interfaces.
```
Comment (carlossanlop) [inline xml/System.IO/WindowsRuntimeStorageExtensions.xml]:
```suggestion
With the Windows 10 Anniversary Update, new interfaces were added to <xref:Windows.Storage.IStorageFolder> and <xref:Windows.Storage.IStorageFile> that allow creating a standard Win32 file handle: `IStorageFolderHandleAccess` and `IStorageItemHandleAccess`. The `CreateSafeFileHandle` group of extension methods take advantage of the `Create` methods of these interfaces.
```

File: xml/System.IO/WindowsRuntimeStorageExtensions.xml [hunk 5]
```diff
@@ -52,6 +52,29 @@ The next example shows the XAML code that is associated with the previous exampl
...
+        <summary>Creates and returns a reference to a safe file handle for the IStorage instance that is being extended.</summary>
+        <remarks>
+          <format type="text/markdown">
+            <![CDATA[
+
+## Remarks
+
+With the Windows 10 Anniversary Update, new interfaces were added to <xref:Windows.Storage.IStorageFolder> and <xref:Windows.Storage.IStorageFile> that allow creating a standard Win32 file handle: `IStorageFolderHandleAccess` and `IStorageItemHandleAccess`. The `CreateSafeFileHandle` group of extension methods take advantage of the `Create` methods of these interfaces.
+
+Getting the Win32 handle allows richer interoperability with both .NET and other existing libraries. You can, for example, go straight to a <xref:System.IO.FileStream> without complicated and potentially performance-sapping wrappers.
```
Comment (JeremyKuhne) [inline xml/System.IO/WindowsRuntimeStorageExtensions.xml]: These sentences aren't really true. The key here is FileStream access to all files. We removed all of the "expensive" wrapping which is why access is problem. With the wrappers there was also a threading issue, but that is gone too. I'd stay away from this whole statement.

File: xml/System.IO/WindowsRuntimeStorageExtensions.xml [hunk 6]
```diff
@@ -52,6 +52,29 @@ The next example shows the XAML code that is associated with the previous exampl
...
+The key thing here is that you can get a brokered `SafeFileHandle`. This is critically important if you want to create a <xref:System.IO.FileStream> around an <xref:Windows.Storage.IStorageItem> or <xref:Windows.Storage.IStorageFolder> when you are running in an [AppContainer](https://docs.microsoft.com/en-us/windows/win32/secauthz/appcontainer-isolation). When running in an AppContainer, the application has very limited file access rights. Things like picture and document folders require the native file handle to be proxied via `RuntimeBroker.exe`, as the application process itself doesn't have rights to most files. This API gets the proxied handle. If accessing user data folders or using a file picker, this API should always be used to create <xref:System.IO.FileStream>. Using a path will often throw <xref:System.AccessViolationException> in these cases.
```
Comment (JeremyKuhne) [inline xml/System.IO/WindowsRuntimeStorageExtensions.xml]: "thing" would be better written as "benefit".  "The key benefit of the API is access to a brokered `SafeFileHandle`."

Comment (carlossanlop) [reply xml/System.IO/WindowsRuntimeStorageExtensions.xml]: Thanks for the suggestion. I am changing it as suggested:

```suggestion
The key benefit of the API is access to a brokered `SafeFileHandle`. This is critically important if you want to create a <xref:System.IO.FileStream> around an <xref:Windows.Storage.IStorageItem> or <xref:Windows.Storage.IStorageFolder> when you are running in an [AppContainer](https://docs.microsoft.com/en-us/windows/win32/secauthz/appcontainer-isolation). When running in an AppContainer, the application has very limited file access rights. Things like picture and document folders require the native file handle to be proxied via `RuntimeBroker.exe`, as the application process itself doesn't have rights to most files. This API gets the proxied handle. If accessing user data folders or using a file picker, this API should always be used to create <xref:System.IO.FileStream>. Using a path will often throw <xref:System.AccessViolationException> in these cases.
```

File: xml/System.IO/WindowsRuntimeStorageExtensions.xml (hunk 5 above)
Comment (carlossanlop) [reply xml/System.IO/WindowsRuntimeStorageExtensions.xml]: Thank you for noticing. I'll remove the two sentences.

**Comment by mairaw** [comment]
Good to merge @JeremyKuhne?

**Comment by carlossanlop** [comment]
@JeremyKuhne do my last changes look good to you? We would like your confirmation so we can get this PR merged. It'll only take a minute.

**Pull request merged**